protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java

```diff
@@ -172,100 +172,136 @@ public void packetReceived(Session session, Packet packet) {
             if (packet instanceof ServerboundKeepAlivePacket keepAlivePacket) {
-                if (keepAlivePacket.getPingId() == this.lastPingId) {
-                    long time = System.currentTimeMillis() - this.lastPingTime;
-                    session.setFlag(MinecraftConstants.PING_KEY, time);
+                KeepAliveState currentKeepAliveState = this.keepAliveState;
+                if (currentKeepAliveState != null) {
+                    if (currentKeepAliveState.keepAlivePending && keepAlivePacket.getPingId() == currentKeepAliveState.keepAliveChallenge) {
+                        currentKeepAliveState.keepAlivePending = false;
+                        session.setFlag(MinecraftConstants.PING_KEY, System.currentTimeMillis() - currentKeepAliveState.keepAliveTime);
+                    } else {
+                        session.disconnect(Component.translatable("disconnect.timeout"));
+                    }
                 }
             } else if (packet instanceof ServerboundConfigurationAcknowledgedPacket) {
-                protocol.setState(ProtocolState.CONFIGURATION);
+                // The developer who sends ClientboundStartConfigurationPacket needs to setOutboundState to CONFIGURATION
+                // after sending the packet. We can't do it in this class because it needs to be a method call right after it was sent.
+                // Using nettys event loop to change outgoing state may cause differences to vanilla.
+                session.switchInboundState(() -> protocol.setInboundState(ProtocolState.CONFIGURATION));
+                keepAliveState = new KeepAliveState();
             } else if (packet instanceof ServerboundPingRequestPacket pingRequestPacket) {
                 session.send(new ClientboundPongResponsePacket(pingRequestPacket.getPingTime()));
                 session.disconnect(Component.translatable("multiplayer.status.request_handled"));
             }
-        } else if (protocol.getState() == ProtocolState.CONFIGURATION) {
-            if (packet instanceof ServerboundFinishConfigurationPacket) {
-                protocol.setState(ProtocolState.GAME);
+        } else if (protocol.getInboundState() == ProtocolState.CONFIGURATION) {
+            if (packet instanceof ServerboundKeepAlivePacket keepAlivePacket) {
+                KeepAliveState currentKeepAliveState = this.keepAliveState;
+                if (currentKeepAliveState != null) {
+                    if (currentKeepAliveState.keepAlivePending && keepAlivePacket.getPingId() == currentKeepAliveState.keepAliveChallenge) {
+                        currentKeepAliveState.keepAlivePending = false;
+                        session.setFlag(MinecraftConstants.PING_KEY, System.currentTimeMillis() - currentKeepAliveState.keepAliveTime);
+                    } else {
+                        session.disconnect(Component.translatable("disconnect.timeout"));
+                    }
+                }
```

Comment:
This code is fairly long and shared between the two states, would be good to move to a method

Reply:
I'd target a rewrite that more closely resembles mojangs impl where each protocol state listener extends a common listener with common logic. I'd not worry too much about duplicated code rn.

Reply:
done